File: python/scenario/_tracing/filters.py
"""Filter presets for controlling which spans are exported to LangWatch."""

from typing import Callable, List

from opentelemetry.sdk.trace import ReadableSpan

# Type alias
SpanFilter = Callable[[ReadableSpan], bool]


def _get_scope_name(span: ReadableSpan) -> str:
    """Get the instrumentation scope name from a span."""
    scope = getattr(span, "instrumentation_scope", None) or getattr(
        span, "instrumentation_info", None
    )
    if scope:
        return getattr(scope, "name", "unknown")
    return "unknown"


def scenario_only(span: ReadableSpan) -> bool:
    """Only keep spans from the scenario instrumentation scope.

    Use this to prevent unrelated server spans (HTTP, middleware, etc.)
    from being exported.

    Example:
        from scenario import setup_scenario_tracing, scenario_only

        setup_scenario_tracing(
            span_filter=scenario_only,
            instrumentors=[],
        )
    """
    return _get_scope_name(span) == "langwatch"


def with_custom_scopes(*scopes: str) -> SpanFilter:
    """Keep spans from scenario scope plus additional custom scopes.

    Example:
        from scenario import setup_scenario_tracing, with_custom_scopes

        setup_scenario_tracing(
            span_filter=with_custom_scopes("my-app/database", "my-app/agent"),
            instrumentors=[],
        )
    """
    allowed = {"langwatch", *scopes}

    def filter_fn(span: ReadableSpan) -> bool:
        return _get_scope_name(span) in allowed

    return filter_fn

File: javascript/src/tracing/filters.ts
/**
 * Criteria for matching spans by instrumentation scope name or span name.
 * Within each field, matchers use OR semantics (any match succeeds).
 * Across fields, AND semantics apply (all specified fields must match).
 */
export interface TraceFilterCriteria {
  instrumentationScopeName?: TraceFilterMatch[];
  name?: TraceFilterMatch[];
}

/**
 * A single match rule for string comparison.
 */
export interface TraceFilterMatch {
  equals?: string;
  startsWith?: string;
  matches?: RegExp;
  ignoreCase?: boolean;
}

/**
 * A filter rule for controlling which spans are exported.
 *
 * Compatible with the langwatch SDK's `TraceFilter` type used by
 * `LangWatchTraceExporter`.
 */
export type TraceFilter =
  | { preset: "vercelAIOnly" | "excludeHttpRequests" }
  | { include: TraceFilterCriteria }
  | { exclude: TraceFilterCriteria };

/**
 * Preset filter that only keeps spans from the @langwatch/scenario instrumentation scope.
 * Use this to prevent unrelated server spans (HTTP, middleware, etc.) from being exported.
 *
 * @example
 * ```typescript
 * import { defineConfig, scenarioOnly } from "@langwatch/scenario";
 * import { LangWatchTraceExporter } from "langwatch/observability";
 *
 * export default defineConfig({
 *   observability: {
 *     traceExporter: new LangWatchTraceExporter({
 *       filters: scenarioOnly,
 *     }),
 *     instrumentations: [], // disable auto-instrumentation
 *   },
 * });
 * ```
 */
export const scenarioOnly: TraceFilter[] = [
  {
    include: {
      instrumentationScopeName: [{ equals: "@langwatch/scenario" }],
    },
  },
];

/**
 * Creates a filter that keeps spans from the @langwatch/scenario scope
 * plus any additional custom instrumentation scopes.
 *
 * @param scopes - Additional instrumentation scope names to include
 * @returns Array of TraceFilter rules
 *
 * @example
 * ```typescript
 * import { defineConfig, withCustomScopes } from "@langwatch/scenario";
 * import { LangWatchTraceExporter } from "langwatch/observability";
 *
 * export default defineConfig({
 *   observability: {
 *     traceExporter: new LangWatchTraceExporter({
 *       filters: withCustomScopes("my-app/database", "my-app/agent"),
 *     }),
 *     instrumentations: [], // disable auto-instrumentation
 *   },
 * });
 * ```
 */
export function withCustomScopes(...scopes: string[]): TraceFilter[] {
  return [
    {
      include: {
        instrumentationScopeName: [
          { equals: "@langwatch/scenario" },
          ...scopes.map((scope) => ({ equals: scope })),
        ],
      },
    },
  ];
}
